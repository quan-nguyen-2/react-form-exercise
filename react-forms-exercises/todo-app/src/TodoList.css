.TodoList {
    display: flex;
    flex-direction: column;
}

.TodoList-form {
    margin: 2rem;
}