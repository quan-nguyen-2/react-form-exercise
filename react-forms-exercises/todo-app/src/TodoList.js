import React, { useState } from "react";
import Todo from "./Todo";
import NewTodoForm from "./NewTodoForm";
import "./TodoList.css";

const TodoList = () => {
  const [todos, setTodos] = useState([]);

  //add a todo to todos array
  const add = (todoObj) => {
    setTodos((todos) => [...todos, todoObj]);
  };

  //update task of a todo using id
  const update = (id, updatedTask) => {
    setTodos((todos) =>
      todos.map((todo) =>
        todo.id === id ? { ...todo, task: updatedTask } : todo
      )
    );
  };

  //remove a todo by id
  const remove = (id) => {
    setTodos((todos) => todos.filter((todo) => todo.id !== id));
  };

  const todoComponents = todos.map((todo) => (
    <Todo
      key={todo.id}
      id={todo.id}
      task={todo.task}
      update={update}
      remove={remove}
    />
  ));

  return (
    <div className="TodoList">
      <div className="TodoList-form">
        <NewTodoForm addTodo={add} />
      </div>
      <ul>{todoComponents}</ul>
    </div>
  );
};

export default TodoList;
