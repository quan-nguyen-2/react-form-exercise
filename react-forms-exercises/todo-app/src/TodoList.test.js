import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import TodoList from "./TodoList";

function addTodo(todoList, task = "Feed chickens!") {
  const taskInput = screen.queryByLabelText("Task");
  fireEvent.change(taskInput, { target: { value: task } });
  const submitBtn = screen.queryByText("Add Todo");
  fireEvent.click(submitBtn);
}

describe("<TodoList /> rendering", function () {
  it("renders without crashing", function () {
    render(<TodoList />);
  });

  it("matches snapshot", function () {
    const { asFragment } = render(<TodoList />);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe("TodoList functionality", function () {
  it("can add a new todo", function () {
    const view = render(<TodoList />);

    //expect todolist to be empty by checking for any delete buttons on page
    expect(screen.queryByText("X")).not.toBeInTheDocument();

    //add a todo and check document for newly added todo
    addTodo(view);
    const newTodo = screen.getByText("Feed chickens!");
    expect(newTodo).toBeInTheDocument();
    expect(screen.getByText("Edit")).toBeInTheDocument();
    expect(screen.getByText("âœ“")).toBeInTheDocument();
    expect(screen.getByText("X")).toBeInTheDocument();

    //expect form input to be empty after submission
    const taskInput = screen.queryByLabelText("Task");
    expect(taskInput.value).toBe("");
  });

  it("can edit a todo", function () {
    const view = render(<TodoList />);
    addTodo(view);

    //click the edit button
    const editBtn = screen.getByText("Edit");
    fireEvent.click(editBtn);
    //change input value from "Feed chickens!" to "Wash the miata"
    const editInput = screen.getByDisplayValue("Feed chickens!");
    fireEvent.change(editInput, { target: { value: "Wash the miata" } });
    //click the update button
    const updateBtn = screen.getByText("Update");
    fireEvent.click(updateBtn);
    //expect "Wash the miata" to be in the document now
    expect(screen.getByText("Wash the miata")).toBeInTheDocument();
  });

  it("can remove a todo", function () {
    addTodo(render(<TodoList />));

    //check for todo in document
    const todo = screen.getByText("Feed chickens!");
    expect(todo).toBeInTheDocument();

    //click on delete button
    const deleteBtn = screen.queryByText("X");
    fireEvent.click(deleteBtn);

    //expect todo to not be in document anymore
    expect(todo).not.toBeInTheDocument();
  });
});
