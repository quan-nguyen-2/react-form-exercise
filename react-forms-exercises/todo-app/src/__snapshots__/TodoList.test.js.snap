// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoList /> rendering matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="TodoList"
  >
    <div
      class="TodoList-form"
    >
      <form>
        <label
          for="task"
        >
          Task
        </label>
        <input
          id="task"
          name="task"
          type="text"
          value=""
        />
        <button>
          Add Todo
        </button>
      </form>
    </div>
    <ul />
  </div>
</DocumentFragment>
`;
